memtable.cc

```diff
@@ -686,23 +686,26 @@ memtable::make_flat_reader(schema_ptr s,
         rd.upgrade_schema(s);
         return rd;
     } else {
-        auto res = make_flat_mutation_reader<scanning_reader>(std::move(s), shared_from_this(), std::move(permit), range, slice, pc, fwd_mr);
+        flat_mutation_reader_v2 res = make_flat_mutation_reader_v2<scanning_reader>(std::move(s), shared_from_this(), std::move(permit), range, slice, pc, fwd_mr);
+        auto res_v1 = downgrade_to_v1(std::move(res));
         if (fwd == streamed_mutation::forwarding::yes) {
-            return make_forwardable(std::move(res));
+            return make_forwardable(std::move(res_v1));
         } else {
-            return res;
+            return res_v1;
```

Comment:
@tgrabiec should we add `flat_mutation_reader_v2::operator flat_mutation_reader() &&` and its v1 equivalent? This would eliminate such glue code and will allow the signatures to change.

Reply:
In fact, if we do that, this patch just becomes a signature change. Anything that doesn't deal directly with the data structures needs minimal adjustments to the type names, not upgrading/downgrading.

Reply:
Such an operator is not a seamless change, because then, overload resolution cannot work with mutation_source:
```cpp
./mutation_reader.hh:236:11: error: call to constructor of 'mutation_source' is ambiguous
        : mutation_source([fn = std::move(fn)] (schema_ptr s,
          ^               ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
./mutation_reader.hh:188:5: note: candidate constructor
    mutation_source(flat_reader_factory_type fn, std::function<partition_presence_checker()> pcf = [] { return make_default_partition_presence_checker(); })
    ^
./mutation_reader.hh:193:5: note: candidate constructor
    mutation_source(flat_reader_v2_factory_type fn, std::function<partition_presence_checker()> pcf = [] { return make_default_partition_presence_checker(); })
    ^
```

I could fix it by hand, but it turns out there are quite a lot of usages.

Reply:
Maybe we can tighten it with constraints? So the constructor at 188 will only accept v1 and the one at 193 will only accept v2 producing functions.

Maybe we'll need to wrap std::function<> in classes (flat_reader_factory_type and flat_reader_v2_factory_type) to have a place to apply the constraints.

Reply:
Even if I do that, nobody passes `std::function<>` to this constructor, everyone uses some lambdas that are implicitly convertible. So what I need to do is to allow converting from the lambda to specific function type, but this conversion is in itself ambiguous. 
Even if I succeed, there are still needed changes, as there is no conversion from flat_mutation_reader to optimized_optional<flat_mutation_reader_v2>, which can be solved by modifying optional's `operator=` from
```
    std::enable_if_t<std::is_same<std::decay_t<U>, T>::value, optimized_optional&>
```
to
```
    std::enable_if_t<std::is_convertible<std::decay_t<U>, T>::value, optimized_optional&>
```

I don't know if such a change is desirable (although Scylla compiles with such a change).

All in all, given additional complexity in understanding the code, I would vouch for not introducing such automatic conversion.

Reply:
@enedil I'm talking about C++20 concepts for constraints, not enable_if (which should be exorcised away).

The motivation for automatic conversion is:
1. reduce boilerplate patches
2. prevent cases where we have a consumer that wants v1 and a producer that produces v1, but has an upgrade_to_v2 at the end. By having automatic conversions, if the producer and consumer match then no conversions happen. If they mismatch then a conversion is automatically inserted.

As an example, after this patch we have exactly this case, the scanning and flushing readers insert a conversion, but their consumers will need the inverse conversion. I'm willing to tolerate a temporary performance loss while we do the conversion process, but here the performance goes purely for conversions that cancel each other out.

Reply:
+1 to not bothering with automatic conversion. When the project is done, the flat_mutation_reader_v1 will be gone anyway so in the long term this is meaningless. It's going to be easier to find and fix all the places that use converters when the conversion will be explicit.

Reply:
Oh, I see you're talking about an existing enable_if. Let's convert it to a concept (without changing the semantics yet), I don't want to see an enable_if ever again.

Reply:
The existing enable_if is a part of Seastar, which still supports C++17.

Reply:
The conversions cancel each other out because the consumers of these readers aren't still expecting f_m_r_v2. Once v1 goes away, there will be no useless conversions.

Reply:
We can introduce flat_mutation_reader_opt and flat_mutation_reader_v2_opt as our own classes, if necessary.

Of course, when everything is v2, we won't have unnecessary conversions. But it will take time to convert everything, and we won't hold up releases for that. I'd like to minimize the performance impact while the conversions are underway.

Reply:
Makes sense @avikivity. Can addition of such conversions be done in a separate PR?

Reply:
Sure

Reply:
@avikivity I don't see how you can avoid conversions with implicit conversions. The conversions will still be there due to a mismatch between the reader impl version and consumer expectation. It's the mismatch which causes conversion, regardless of whether the conversion is implicit or explicit.

Reply:
If you have a producer doing downgrade_to_v1() and a consumer doing upgrade_to_v2(), then the conversions would be eliminated, if both go via mutation_source. 

But I see here the situation is more complex, there are two branches of an if that internally use v1 and v2, and the result is not a mutation_source but must be one of the two. So here no games will prevent the conversion.